File: 0070-climbing-stairs/0070-climbing-stairs.py
class Solution:
    def climbStairs(self, n: int) -> int:
        memo = [0]*(n+1)
        def helper(i, n):
            if i>n:
                return 0
            if i==n:
                return 1
            if memo[i]:
                return memo[i]
            memo[i] = helper(i+1, n) + helper(i+2,n)
            return memo[i]
            
        return helper(0, n)


'''
                0
        1               2
    1       2       1       2
1   2.    1.  2.  1.  2.  1   2
n=3
    0 0 0 0
'''        
        

File: 0070-climbing-stairs/0070-climbing-stairs.cpp
class Solution {
public:
    int climbStairs(int n) {
        int stairs[n+1];

        stairs[0] = 1;
        stairs[1] = 1;

        for(int i=2; i<=n; i++) {
            stairs[i] = stairs[i-1] + stairs[i-2];
        }
        return stairs[n];
    }
};

/*
n
       1          2
    1     2    1.     2
 1.   2. 1. 2 1  2. 1.  2

 stairs[n+1] = {0,}
 stairs[0] = 1
 stairs[1] = 1
 stairs[2] = 2
 stairs[3] = 3

 stair[n] = stairs[n-1] + stairs[n-2]
*/